lightning/src/chain/channelmonitor.rs

```diff
@@ -3111,10 +3111,21 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 		F::Target: FeeEstimator,
 		L::Target: Logger,
 	{
-		self.onchain_events_awaiting_threshold_conf.retain(|ref entry| if entry.txid == *txid {
-			log_info!(logger, "Removing onchain event with txid {}", txid);
-			false
-		} else { true });
+		let mut height = None;
+		for entry in self.onchain_events_awaiting_threshold_conf.iter() {
+			if entry.txid == *txid {
+				height = Some(entry.height);
+				break;
+			}
+		}
+
+		if let Some(height) = height {
+			log_info!(logger, "transaction_unconfirmed of txid {} implies height {} was reorg'd out", txid, height);
+			self.onchain_events_awaiting_threshold_conf.retain(|ref entry| entry.height < height);
+		}
+
+		debug_assert!(!self.onchain_events_awaiting_threshold_conf.iter().any(|ref entry| entry.txid == *txid));
+
 		self.onchain_tx_handler.transaction_unconfirmed(txid, broadcaster, fee_estimator, logger);
```

Comment:
Well all the `OnchainEventEntry` do have a `txid` we could go batch the call to `transaction_unconfirmed()` ?